backend/functions/tasks.py

```diff
@@ -204,4 +203,161 @@ def sentence_text_translate_and_save_translation_pairs(
+
+
+@shared_task(bind=True)
+def conversation_data_machine_translation(
+    self,
+    languages,
+    input_dataset_instance_id,
+    output_dataset_instance_id,
+    batch_size,
+    api_type,
+    checks_for_particular_languages,
+):
+    """Function to translate Conversation data item and to save the translations in another Conversation dataitem.
+
+    Args:
+        languages (list): List of output languages for the translations.
+        input_dataset_instance_id (int): ID of the input dataset instance.
+        output_dataset_instance_id (int): ID of the output dataset instance.
+        batch_size (int): Number of sentences to be translated in a single batch.
+        api_type (str): Type of API to be used for translation. (default: indic-trans)
+            Allowed - [indic-trans, google]
+        checks_for_particular_languages (bool): If True, checks for the particular languages in the translations.
+    """
+
+    # Get the output dataset instance
+    output_dataset_instance = dataset_models.DatasetInstance.objects.get(
+        instance_id=output_dataset_instance_id
+    )
+
+    # Collect all the Conversation dataitems for the input DatasetInstance and convert to dataframe
+    conversation_dataitems = dataset_models.Conversation.objects.filter(
+        instance_id=input_dataset_instance_id
+    ).values_list("id", "scenario", "prompt", "conversation_json", "language")
+
+    conversation_dataitems_df = pd.DataFrame(
+        conversation_dataitems,
+        columns=["id", "scenario", "prompt", "conversation_json", "language"],
+    )
+
+    # Check if the dataframe is empty
+    if conversation_dataitems_df.shape[0] == 0:
+
+        # Update the task status
+        self.update_state(
+            state="FAILURE",
+            meta={
+                "No sentences to upload.",
+            },
+        )
+
+        raise Exception("The conversation data is empty.")
+
+    # Iterate through the languages
+    for output_language in languages:
+
+        all_translated_conversation_objects = []
+        # Iterate through the conversation dataitems
+        for index, row in conversation_dataitems_df.iterrows():
+
+            # Get the instance of the Conversation dataitem
+            conversation_dataitem = dataset_models.Conversation.objects.get(
+                id=row["id"]
+            )
+
+            # Get the conversation JSON and iterate through it
+            conversation_json = row["conversation_json"]
+            translated_conversation_json = []
+            for conversation in conversation_json:
+
+                # Get the sentence list, scenario and prompt
+                sentences_to_translate = dict(conversation).get("sentences", [])
+                speaker_id = dict(conversation).get("speaker_id")
+                sentence_count = len(sentences_to_translate)
+                sentences_to_translate.append(row["scenario"])
+                sentences_to_translate.append(row["prompt"])

```

Comment:
Isn't it redundant to translate this each and everytime?
(Because scenario and prompt are going to remain the same for a given dataitem.

Shall we just translate it at the last?

Reply:
Updated in #441